models/base.py

```diff
@@ -1243,23 +1247,39 @@ def remove_data_from_disk(self, remove_folders=True):
         remove_folders: bool
             If True, will remove any folders on the path to the files
             associated to this object, if they are empty.
+        remove_downstream_data: bool
+            If True, will also remove any downstream data.
+            Will recursively call get_downstreams() and find any objects
+            that have remove_data_from_disk() implemented, and call it.
+            Default is False.
         """
-        if self.filepath is None:
-            return
-        # get the filepath, but don't check if the file exists!
-        for f in self.get_fullpath(as_list=True, nofile=True):
-            if os.path.exists(f):
-                os.remove(f)
-                if remove_folders:
-                    folder = f
-                    for i in range(10):
-                        folder = os.path.dirname(folder)
-                        if len(os.listdir(folder)) == 0:
-                            os.rmdir(folder)
-                        else:
-                            break
+        if self.filepath is not None:
+            # get the filepath, but don't check if the file exists!
+            for f in self.get_fullpath(as_list=True, nofile=True):
+                if os.path.exists(f):
+                    os.remove(f)
+                    if remove_folders:
+                        folder = f
+                        for i in range(10):
+                            folder = os.path.dirname(folder)
+                            if len(os.listdir(folder)) == 0:
+                                os.rmdir(folder)
+                            else:
+                                break
+
+        if remove_downstream_data:
+            try:
+                downstreams = self.get_downstreams()
+                for d in downstreams:
+                    if hasattr(d, 'remove_data_from_disk'):
+                        d.remove_data_from_disk(remove_folders=remove_folders, remove_downstream_data=True)
 
-    def delete_from_disk_and_database(self, session=None, commit=True, remove_folders=True):
+            except NotImplementedError as e:
+                pass  # if this object does not implement get downstreams, it is ok
+
+    def delete_from_disk_and_database(
```

Comment:
Is there a test that verifies that when this is called with remove_downstream_data=True, the downstreams get deleted from disk, database, and archive?  (If not, we should write one; make an Issue.)

Reply:
I made a new issue.